### PR TITLE
[Quantization] Add symmetric with power2 scale quantization schema

### PR DESCRIPTION
**Pull request by mciprian13**
**Summary** 
Add new quantization schema (named "SymmetricWithPower2Scale").
The new quantization schema is a particular quantization schema which:
- is symmetric (offset = 0)
- the floating point scale if forced to be a power of 2 (**scale = 2 ^ exp** for example 0.5, 0.125, etc)

For the previous quantization schemes the main approximation was:
- (x * float_scale + offset) -> ((x >> pre_scale  * integer_scale) >> post_scale) + offset

For the new quantization schema the approximation has a particular, more simpler form:
- (x * float_scale + offset) -> x * integer_scale for positive exponent
- (x * float_scale + offset) -> x >> post_scale for negative exponent

The new quantization schema:
- is expected to be less accurate (in terms of arithmetic approximation of floating-point operations)
- is expected to be simpler (arithmetic complexity, number of operations)
- can be viewed as another point in the spectrum of quantization schemes with different tradeoffs between accuracy and complexity (this is oriented more towards the "simple"/"less accurate" end)
- has proved to be good enough in terms of accuracy for some networks (e.g. lenet_mnist, cifar10)
where the network was sufficiently robust/redundant (1..2 % accuracy drop relative to floating-point)
- the main point is that it allows the specialization of some the LIBJIT kernels with other open-source hand-optimized kernels/libraries for some specific architectures

This quantization schema requires the floating-point scales to preserve their bit-exact values during YAML round-trip. Therefore, I added a workaround for retrieving the floating-point scales bit-exactly during YAML round-trip.

**Documentation**
None.

**Test Plan**
Tests for new quantization schema.
Tests for verifying that floating-point numbers are preserved bit-exactly during YAML round-trip.

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

**Comment by rdzhabarov** [comment]
Thanks @mciprian13 for contribution! This schema could be beneficial for multiple vendor.

Can you sign the CLA and make sure tests are passing.

File: lib/Quantization/Base/Base.cpp [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /**
  * Copyright (c) 2017-present, Facebook, Inc.
+ * Copyright (c) 2019-present, NXP Semiconductor, Inc.
```
Comment (rdzhabarov) [inline lib/Quantization/Base/Base.cpp]: this should not be here based on the CLA agreement.

Comment (mciprian13) [reply lib/Quantization/Base/Base.cpp]: I understand. I removed it.

File: include/glow/Quantization/Base/Base.h [hunk 2]
```diff
@@ -53,7 +53,7 @@ struct QuantizationTransform32To8 {
   int32_t transform(int32_t input) {
     // The operation x >> y is rounded down to negative infinity. To get to
     // round-nearest we add (1 << (shift - 1)) to the value prior to shifting.
-    int rtn = (1 << (post - 1));
+    int rtn = (post > 0) ? (1 << (post - 1)) : 0;
```
Comment (beicy) [inline include/glow/Quantization/Base/Base.h]: Please update the comments since "rtn" is updated.

File: lib/Quantization/Base/Base.cpp [hunk 3]
```diff
@@ -201,6 +201,27 @@ QuantizationTransform32To8 quantizeScaleOffset32To8(float scale,
...
   int postShift = 0;
 
+  // We treat first the particular case when scale is a power of 2 (2 ^ exp,
+  // where exp is a signed integer exponent). The operation is specialized as:
+  // - for positive 2's exponent:
+  //     x * scale + offset (pre = 0, post = 0, scale = (int)scale).
+  // - for negative 2's exponent:
+  //     x >> post + offset (pre = 0, post = -exp, scale = 1).
+  if (isFloatPowerOf2(scale)) {
+    int exp = getFloat2Exp(scale);
```
Comment (beicy) [inline lib/Quantization/Base/Base.cpp]: Just want to double check, if "SymmetricWithPower2Scale" is not selected, should this flow be executed? Thanks!

Comment (mciprian13) [reply lib/Quantization/Base/Base.cpp]: Yes.
- When the scale is a power of 2 (whether it was forced using the "SymmetricWithPower2Scale" schema or obtained by coincidence for other schema) it always makes sense to approximate the multiplication using only one operation (up/down scale).
- The previous logic, for example, for a scale = (exactly) 0.125, it yielded the following expression: x * 0.125 -> (x * 256) >> 11 (integer_scale = 256 and postShift=11), whereas it makes more sense to do just x >> 3 (integer_scale = 1 and postShift = 3).

As a side note, I don`t think the logic of this function (quantizeScaleOffset32To8) is rigorous enough (mathematically). Choosing the pre/post/scale should also take into account: data dynamic range in the quantized domain, the bit-width of the intermediate computation (accumulator), etc. But this will be the subject of a future PR.

Thanks!

**Comment by mciprian13** [comment]
It seems the tests are failing for other reasons than my code updates (some connection issues). Can you please verify (or rerun the tests if applicable)?
Thanks!

**Comment by rdzhabarov** [comment]
Seems like everything is passing now. I'll get this reviewed today.

**Comment by beicy** [comment]
Please update the comments about "int rtn = (post > 0) ? (1 << (post - 1)) : 0;" . After that, I think it is good to merge :) Thanks!

File: include/glow/Quantization/Base/Base.h (hunk 2 above)
Comment (mciprian13) [reply include/glow/Quantization/Base/Base.h]: Done.

File: lib/Quantization/Serialization.cpp [hunk 4]
```diff
@@ -28,11 +28,52 @@
...
+  static QuotingType mustQuote(StringRef) { return QuotingType::None; }
+};
+
 /// Mapping for NodeQuantizationInfo yaml serializer.
 template <> struct MappingTraits<glow::NodeQuantizationInfo> {
+  struct FloatNormalized {
+    FloatNormalized(IO &io) : _val(0.0) {}
+    FloatNormalized(IO &, float &val) : _val(val) {}
+    float denormalize(IO &) { return _val._val; }
+    FloatWrapper _val;
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Serialization.cpp]: Changed.

File: lib/Quantization/Serialization.cpp [hunk 5]
```diff
@@ -28,11 +28,52 @@
...
+/// The default behavior of YAML is to serialize floating point numbers
+/// using the "%g" format specifier which is not guaranteed to print all
+/// the decimals. During a round-trip (serialize, deserialize) decimals
+/// might be lost and hence precision is lost. Although this might not be
+/// critical for some quantization schema, for "SymmetricWithPower2Scale"
+/// the round-trip must preserve the exact representation of the floating
+/// point scale which is a power of 2. The code below is a workaround to
+/// overwrite the behavior of the YAML serializer to print all the digits.
+struct FloatWrapper {
+  float _val;
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Serialization.cpp]: Changed.

File: lib/Quantization/Serialization.cpp [hunk 6]
```diff
@@ -28,11 +28,52 @@
...
+  FloatWrapper(float val) : _val(val) {}
+};
+
+template <> struct ScalarTraits<FloatWrapper> {
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Serialization.cpp]: I tried in the past to do that but I did not succeed. Can`t remember the details.

File: tools/loader/Loader.cpp [hunk 7]
```diff
@@ -85,7 +85,10 @@ llvm::cl::opt<quantization::Schema> quantizationSchema(
                    "Use symmetric ranges"),
         clEnumValN(quantization::Schema::SymmetricWithUnsigned,
                    "symmetric_with_uint8",
-                   "Use symmetric ranges with potentially uint8 ranges")),
+                   "Use symmetric ranges with potentially uint8 ranges"),
+        clEnumValN(quantization::Schema::SymmetricWithPower2Scale,
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply tools/loader/Loader.cpp]: I updated the doc to mention the new schema.

File: lib/Quantization/Base/Base.cpp [hunk 8]
```diff
@@ -338,6 +360,11 @@ TensorQuantizationParams chooseQuantizationParams(float min, float max,
     nudgedZeroPoint = static_cast<int32_t>(round(initialZeroPoint));
   }
 
+  // For SymmetricWithPower2Scale, round scale to nearest higher power of 2.
+  if (schema == quantization::Schema::SymmetricWithPower2Scale) {
+    scale = std::exp2(std::ceil(std::log2(scale)));
```
[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Base/Base.cpp]: There is an assert in the same function of code which does that all the time.
I did however add an unit test to make the verification: chooseQuantizationSymmetricWithPower2Scale

[the remark this answers is not in the document]
Comment (mciprian13) [reply lib/Quantization/Base/Base.cpp]: While adding the last unit test I found a bug (or corner case) in the function "chooseQuantizationParams":
- for the previous "Symmetric" schema (and also for the newly added "SymmetricWithPower2Scale"), when qTy=UInt8QTy (qmin = 0) we get division by zero in the following line of code:
double rmin = min / (double)qmin;

What should be do about this? This problem was already there and was not exposed by unit tests. Do you think we should exclude this case by putting an assert in the function?

Thanks!

**Comment by rdzhabarov** [comment]
Internal tests are failing with

```
Quantization.chooseQuantizationSymmetricWithPower2Scale
glow/glow/lib/Quantization/Base/Base.cpp:305:23: runtime error: division by zero
    #0 0x7fa30255fc9c in glow::quantization::chooseQuantizationParams(float, float, glow::quantization::Schema, glow::ElemKind) glow/glow/lib/Quantization/Base/Base.cpp:305
    #1 0x36f301 in glow::chooseQuantParamsPower2Scale(float, float, glow::ElemKind) glow/glow/tests/unittests/QuantizationTest.cpp:1370
    #2 0x36ef61 in glow::Quantization_chooseQuantizationSymmetricWithPower2Scale_Test::TestBody() glow/glow/tests/unittests/QuantizationTest.cpp:1378
    #3 0x7fa300b213de in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2473:29
    #4 0x7fa300b1244a in testing::Test::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2490:50
    #5 0x7fa300b1244a in testing::Test::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2480:6
    #6 0x7fa300b125a0 in testing::TestInfo::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2677:14
    #7 0x7fa300b125a0 in testing::TestInfo::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2651:6
    #8 0x7fa300b126b3 in testing::TestCase::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2802:31
    #9 0x7fa300b126b3 in testing::TestCase::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2787:6
    #10 0x7fa300b12dd3 in testing::internal::UnitTestImpl::RunAllTests() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:4739:46
    #11 0x7fa300b13084 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:2473:29
    #12 0x7fa300b13084 in testing::UnitTest::Run() /home/engshare/third-party2/googletest/master/src/googletest/googletest/src/gtest.cc:4333:55
    #13 0x7fa305533430 in RUN_ALL_TESTS() third-party-buck/platform007/build/googletest/include/gtest/gtest.h:2255
    #14 0x7fa305533307 in main common/gtest/LightMain.cpp:19

SUMMARY: UndefinedBehaviorSanitizer: float-divide-by-zero glow/glow/lib/Quantization/Base/Base.cpp:305:23 in 
"/usr/bin/time:
MaxMemory 99316
Inputs 0
Outputs 0"
```

Could you check this?

**Comment by mciprian13** [comment]
Are they still failing? It it was the previous ASAN problem I fixed it.

**Comment by rdzhabarov** [comment]
> Are they still failing? It it was the previous ASAN problem I fixed it.

I'll get it tested and get back.

**Comment by rdzhabarov** [comment]
@mciprian13 required tests are passing, could you resolve conflict on `include/glow/Quantization/Base/Base.h` and i'll get this merged.
